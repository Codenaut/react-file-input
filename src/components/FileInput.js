import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { handleChangeEvent, handleDropEvent, preventDefault } from '../helpers/event';
import { getImageThumbnail, createImageFromSource, updateFileFromBlob } from '../helpers/file';
import { selectIsDragging, selectIsDraggingOver } from '../helpers/fileInputSelectors';
import { IMAGE_MIME_TYPE } from '../helpers/mime';

import DropArea from './DropArea';
import FileInputMetadata from './FileInputMetadata';
import ImageThumbnail from './ImageThumbnail';
import ImageEditor from './ImageEditor';

import CanvasPrinter from './CanvasPrinter';

import '../styles/FileInput.scss';

class FileInput extends Component {
  constructor(props) {
    super(props);

    this.state = {
      enteredInDocument: 0,
      isOver: 0,
      value: null,
      image: null,
      hasBeenEdited: false,
    };

    this.input = null;
    this.canvasPrinter = null;

    this.handleFile = this.handleFile.bind(this);
    this.handleImageFile = this.handleImageFile.bind(this);
    this.cancelEdition = this.cancelEdition.bind(this);
    this.saveEdition = this.saveEdition.bind(this);

    this.openFileDialog = this.openFileDialog.bind(this);
    this.selectFile = this.selectFile.bind(this);

    this.onDocumentDragEnter = this.onDocumentDragEnter.bind(this);
    this.onDocumentDragLeave = this.onDocumentDragLeave.bind(this);
    this.onDocumentDrop = this.onDocumentDrop.bind(this);

    this.onDragEnter = this.onDragEnter.bind(this);
    this.onDragLeave = this.onDragLeave.bind(this);
    this.onDrop = this.onDrop.bind(this);
  }

  componentDidMount() {
    const { dragOnDocument, dropOnDocument } = this.props;

    if (dragOnDocument) {
      document.addEventListener('dragenter', this.onDocumentDragEnter, false);
      document.addEventListener('dragleave', this.onDocumentDragLeave, false);
      document.addEventListener('drop', this.onDocumentDrop, false);
    }

    if (!dropOnDocument) {
      document.addEventListener('dragover', preventDefault, false);
      document.addEventListener('drop', preventDefault, false);
    }
  }

  componentWillUnmount() {
    const { dragOnDocument, dropOnDocument } = this.props;

    if (dragOnDocument) {
      document.removeEventListener('dragenter', this.onDocumentDragEnter);
      document.removeEventListener('dragleave', this.onDocumentDragLeave);
      document.removeEventListener('drop', this.onDocumentDrop);
    }

    if (!dropOnDocument) {
      document.removeEventListener('dragover', preventDefault);
      document.removeEventListener('drop', preventDefault);
    }
  }

  openFileDialog() {
    if (this.input) {
      this.input.click();
    }
  }

  async cancelEdition(image) {
    const { scaleOptions } = this.props;
    if (!scaleOptions) {
      return this.setState({ hasBeenEdited: true });
    }

    const { displayImageThumbnail } = this.props;
    const { value: file } = this.state;

    const resizedImageBlob = await this.canvasPrinter.resizeImage(image, scaleOptions);

    const resizedFile = updateFileFromBlob(resizedImageBlob, file);
    const resizedThumbnail = displayImageThumbnail
      ? await getImageThumbnail(resizedFile)
      : null;


    return this.setState(state => ({
      ...state,
      enteredInDocument: 0,
      isOver: 0,
      value: resizedFile,
      image: resizedThumbnail,
      hasBeenEdited: true,
    }));
  }

  async saveEdition(image, area) {
    const { displayImageThumbnail, scaleOptions } = this.props;
    const { value: file } = this.state;

    const croppedImageBlob = await this.canvasPrinter.cropAndResizeImage(image, area, scaleOptions);

    const croppedFile = updateFileFromBlob(croppedImageBlob, file);
    const croppedThumbnail = displayImageThumbnail
      ? await getImageThumbnail(croppedFile)
      : null;


    return this.setState(state => ({
      ...state,
      enteredInDocument: 0,
      isOver: 0,
      value: croppedFile,
      image: croppedThumbnail,
      hasBeenEdited: true,
    }));
  }

  handleFile(file, callback = null) {
    this.setState(state => ({
      ...state,
      enteredInDocument: 0,
      isOver: 0,
      value: file,
      image: null,
      hasBeenEdited: true,
    }), () => {
      if (callback) {
        callback(this.state);
      }
    });
  }

  async handleImageFile(file, callback = null) {
    const { displayImageThumbnail, scaleOptions, cropTool } = this.props;

    const image = cropTool || scaleOptions || displayImageThumbnail
      ? await getImageThumbnail(file)
      : null;

    if (cropTool || !scaleOptions) {
      return this.setState(state => ({
        ...state,
        enteredInDocument: 0,
        isOver: 0,
        value: file,
        image,
        hasBeenEdited: !cropTool,
      }), () => {
        if (callback) {
          callback(this.state);
        }
      });
    }

    const imageObject = await createImageFromSource(image);
    const resizedImageBlob = await this.canvasPrinter.resizeImage(imageObject, scaleOptions);

    const resizedFile = updateFileFromBlob(resizedImageBlob, file);
    const resizedThumbnail = displayImageThumbnail
      ? await getImageThumbnail(resizedFile)
      : null;


    return this.setState(state => ({
      ...state,
      enteredInDocument: 0,
      isOver: 0,
      value: resizedFile,
      image: resizedThumbnail,
      hasBeenEdited: !cropTool,
    }), () => {
      if (callback) {
        callback(this.state);
      }
    });
  }

  selectFile(event) {
    const { onChangeCallback } = this.props;

    const files = handleChangeEvent(event);

    if (files.length) {
      const file = files[0];

      if (IMAGE_MIME_TYPE.test(file.mimeType)) {
        this.handleImageFile(file, onChangeCallback);
      } else {
        this.handleFile(file, onChangeCallback);
      }
    }
  }

  onDocumentDragEnter(event) {
    this.setState(state => ({ ...state, enteredInDocument: state.enteredInDocument + 1 }));
  }
  onDocumentDragLeave(event) {
    this.setState(state => ({ ...state, enteredInDocument: state.enteredInDocument - 1 }));
  }
  onDocumentDrop(event) {
    this.setState(state => ({ ...state, enteredInDocument: 0 }));
  }

  onDragEnter(event) {
    const { onDragEnterCallback } = this.props;
    const wasDraggingOver = selectIsDraggingOver(this.state);

    this.setState(state => ({ ...state, isOver: state.isOver + 1 }), () => {
      if (onDragEnterCallback && !wasDraggingOver) {
        onDragEnterCallback(this.state);
      }
    });
  }
  onDragLeave(event) {
    const { onDragLeaveCallback } = this.props;
    const wasDraggingOver = selectIsDraggingOver(this.state);

    this.setState(state => ({ ...state, isOver: state.isOver - 1 }), () => {
      if (onDragLeaveCallback && wasDraggingOver) {
        onDragLeaveCallback(this.state);
      }
    });
  }
  onDrop(event) {
    const { onDropCallback } = this.props;

    const files = handleDropEvent(event);

    if (files.length) {
      const file = files[0];

      if (IMAGE_MIME_TYPE.test(file.mimeType)) {
        this.handleImageFile(file, onDropCallback);
      } else {
        this.handleFile(file, onDropCallback);
      }
    }
  }

  render() {
    const { value, image, hasBeenEdited } = this.state;
    const {
      customMetadata: CustomMetadata,
      customImageThumbnail: CustomImageThumbnail,
      label,
      cropTool,
      scaleOptions,
    } = this.props;

    const MetadataClass = CustomMetadata || FileInputMetadata;
    const ImageThumbnailClass = CustomImageThumbnail || ImageThumbnail;

    const isDragging = selectIsDragging(this.state);

    const metadataComponent = value
      && <MetadataClass name={value.filename} size={value.size} extension={value.extension} type={value.mimeType}/>;
    const imageThumbnailComponent = image && <ImageThumbnailClass image={image}/>;

    return (
      <div className="brainhub-file-input__wrapper">
        <div className="brainhub-file-input__label">{label}</div>
        <input
          className="brainhub-file-input__input--hidden"
          type="file"
          ref={ref => {
            this.input = ref;
          }}
          onChange={this.selectFile}
        />
        <div className="brainhub-file-input__fileInfo">
          {this.props.displayImageThumbnail && imageThumbnailComponent}
          {this.props.displayMetadata && metadataComponent}
        </div>
        <DropArea
          dragging={isDragging}
          onDragEnter={this.onDragEnter}
          onDragLeave={this.onDragLeave}
          onDrop={this.onDrop}
          openFileDialog={this.openFileDialog}
        />
        {image && cropTool && !hasBeenEdited
          ? (
            <ImageEditor
              image={image}
              onCancelEdition={this.cancelEdition}
              onSaveEdition={this.saveEdition}
            />
          )
          : null}
        { (cropTool || scaleOptions)
          && (
            <CanvasPrinter
              ref={ref => {
                this.canvasPrinter = ref;
              }}
            />
          )
        }

      </div>
    );
  }
}

FileInput.defaultProps = {
  dragOnDocument: true,
  dropOnDocument: false,
  onChangeCallback: null,
  onDragEnterCallback: null,
  onDragLeaveCallback: null,
  onDropCallback: null,
  displayMetadata: true,
  displayImageThumbnail: true,
  scaleOptions: null,
  cropTool: false,
};

FileInput.propTypes = {
  dragOnDocument: PropTypes.bool,
  dropOnDocument: PropTypes.bool,
  label: PropTypes.string.isRequired,
  onChangeCallback: PropTypes.func,
  onDragEnterCallback: PropTypes.func,
  onDragLeaveCallback: PropTypes.func,
  onDropCallback: PropTypes.func,
  displayMetadata: PropTypes.bool,
  displayImageThumbnail: PropTypes.bool,
  customMetadata: PropTypes.func,
  customImageThumbnail: PropTypes.func,
  scaleOptions: PropTypes.shape({
    width: PropTypes.number,
    height: PropTypes.number,
    keepAspectRatio: PropTypes.boolean,
  }),
  cropTool: PropTypes.bool,
};

export default FileInput;
