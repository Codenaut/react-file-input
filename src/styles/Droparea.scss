@import "global";

.brainhub-droparea {
  height: 90px;
  font-family: $FONT_FAMILY;
  display: flex;
  box-sizing: border-box;
  margin: 4px 0;
}

.brainhub-droparea--dragging {
  border: 1px dashed $LIGHT_GREEN;
}

.brainhub-drpoparea__info {
  flex: 1;
  display: flex;
  justify-content: center;
  flex-direction: column;
  text-align: center;
  color: $LIGHT_GREEN;
}

.brainhub-drpoparea__button {
  font-size: 14px;
  line-height: 18px;
  color: $LIGHT_GREEN;
}