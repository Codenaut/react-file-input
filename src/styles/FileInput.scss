@import 'global';

.brainhub-file-input__wrapper {
  width: 25%;
}

.brainhub-file-input__input--hidden {
  display: none;
}
