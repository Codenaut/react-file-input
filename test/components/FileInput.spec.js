import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import PropTypes from 'prop-types';

import Droparea from '../../src/components/Droparea';
import FileInput from '../../src/components/FileInput';

const defaultProps = {
  label: 'Test',
};

const setup = (props = {}) => {
  const fileInput = mount(<FileInput {...defaultProps} {...props} />);

  return {
    fileInput,
    label: fileInput.find('.brainhub-file-input__label'),
    input: fileInput.find('input[type="file"]'),
    droparea: fileInput.find(Droparea),
  };
};

const CustomComponent = ({ name, size }) => <div>{name}:{size}</div>;
CustomComponent.propTypes = {
  name: PropTypes.string,
  size: PropTypes.number,
};

describe('components', () => {
  describe('FileInput', () => {
    const data = {
      name: 'Cute puppies',
      size: 1000,
    };

    it('should render with a default state', () => {
      const { fileInput } = setup();

      expect(fileInput.state()).toEqual({
        enteredInDocument: 0,
        isOver: 0,
        value: null,
      });
    });

    it('should render a hidden file input', () => {
      const { input } = setup();
      expect(input).toHaveLength(1);
      expect(input.prop('type')).toBe('file');
      expect(input.hasClass('brainhub-file-input__input--hidden')).toBeTruthy();
    });

    it('should render a label', () => {
      const { label } = setup();

      expect(label).toHaveLength(1);
      expect(label.text()).toBe(defaultProps.label);
    });

    it('should render a droparea', () => {
      const { droparea } = setup();

      expect(droparea).toHaveLength(1);
    });

    it('should render a droparea that can open the file dialog', () => {
      const { droparea, input } = setup();

      const inputClickSpy = jest.spyOn(input.instance(), 'click');

      droparea.prop('openFileDialog')();

      expect(inputClickSpy).toHaveBeenCalled();
    });

    it('should call onDragEnter callback when drag enters for first time', () => {
      const onDragEnterCallback = jest.fn();
      const { droparea, fileInput } = setup({ onDragEnterCallback });

      droparea.simulate('dragenter');

      expect(onDragEnterCallback).toHaveBeenCalledWith(fileInput.state());

      onDragEnterCallback.mockClear();
      droparea.simulate('dragenter');

      expect(onDragEnterCallback).not.toHaveBeenCalled();
    });

    it('should decreate `isOver` state when drag leaves', () => {
      const { droparea, fileInput } = setup();

      const state = fileInput.state();

      droparea.simulate('dragLeave');

      expect(fileInput.state()).toEqual({
        ...state,
        isOver: state.isOver - 1,
      });
    });

    it('should call onDragLeaves callback when drag leaves for last time', () => {
      const onDragLeaveCallback = jest.fn();
      const { droparea, fileInput } = setup({ onDragLeaveCallback });

      fileInput.setState({ isOver: 1 });
      droparea.simulate('dragLeave');

      expect(onDragLeaveCallback).toHaveBeenCalledWith(fileInput.state());

      onDragLeaveCallback.mockClear();
      droparea.simulate('dragLeave');

      expect(onDragLeaveCallback).not.toHaveBeenCalled();
    });

    it('should store the drop file in the state, and reset the rest, and call the callback', () => {
      const onDropCallback = jest.fn();
      const { droparea, fileInput } = setup({ onDropCallback });

      const file = { filename: 'MockFile' };
      fileInput.setState({ isOver: 2, enteredInDocument: 2 });

      droparea.simulate('drop', { dataTransfer: { files: [file] } });

      expect(fileInput.state('enteredInDocument')).toBe(0);
      expect(fileInput.state('isOver')).toBe(0);
      expect(fileInput.state('value')).toBe(file);
      expect(onDropCallback).toHaveBeenCalledWith(fileInput.state());
    });

    it('should render metadata from default component', () => {
      const { fileInput } = setup();
      fileInput.setState({ value: data });

      expect(fileInput.find('FileInputMetadata')).toHaveLength(1);
    });

    it('should render metadata from custom component', () => {
      const { fileInput } = setup({ customMetadata: CustomComponent });
      fileInput.setState({ value: data });

      expect(fileInput.find(CustomComponent)).toBeTruthy();
    });

    it('should not render metadata when user pass false to displayMetadata prop', () => {
      const { fileInput } = setup({ displayMetadata: false });
      fileInput.setState({ value: data });

      expect(fileInput.find('FileInputMetadata').length).toEqual(0);
    });

    it('should match exact snapshot', () => {
      const tree = renderer.create(
        <div>
          <FileInput {...defaultProps} />
        </div>
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
