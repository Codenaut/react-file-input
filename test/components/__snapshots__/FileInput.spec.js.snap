// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components FileInput should match exact snapshot 1`] = `
<div>
  <div
    className="brainhub-file-input__wrapper"
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDragOver={undefined}
    onDrop={[Function]}
  >
    <input
      className="brainhub-file-input__input--hidden"
      onChange={[Function]}
      type="file"
    />
    <button
      onClick={[Function]}
    >
      Select File
    </button>
    <div
      className="brainhub-file-input__dropInfo--hidden"
    >
      <p>
        Drop here to select file
      </p>
    </div>
  </div>
</div>
`;

exports[`components FileInputMetadata should match exact snapshot 1`] = `
<div>
  <div
    className="brainhub-file-input__metadata"
  >
    <div
      className="brainhub-file-input__metadata__image"
    >
      <div
        style={
          Object {
            "backgroundColor": "#5680BB",
            "height": "20px",
            "width": "20px",
          }
        }
      />
    </div>
    <div
      className="brainhub-file-input__metadata__info"
    >
      <span
        className="brainhub-file-input__metadata__info_name"
      >
        tes
      </span>
      t
       (
      1000 B
      )
    </div>
  </div>
</div>
`;
